Jenkinsfile

```diff
@@ -165,7 +165,8 @@ rocmtest clang_debug: rocmnode('mi100+') { cmake_build ->
 }, clang_asan: rocmnode('nogpu') { cmake_build ->
     stage('Clang ASAN') {
         def sanitizers = "undefined,address"
-        def debug_flags = "-g -O2 -fno-omit-frame-pointer -fsanitize=${sanitizers} -fno-sanitize-recover=${sanitizers}"
+        def sanitizers_disabled = "float-cast-overflow"
```

Comment:
agree with chris. 

@nives-vukovic  can you try disabling sanitizer just on the convert op's compute function and see if that works ? 
https://clang.llvm.org/docs/AddressSanitizer.html#disabling-instrumentation-with-attribute-no-sanitize-address

Reply:
The issue is reported in AMDMIGraphX/src/include/migraphx/shape.hpp, and when I add `__attribute__((no_sanitize("float-cast-overflow")))` above:
```
 type operator()(U u) const
        {
            return type(u);
        }
```
The issue is not reported on our system. However, this is a common function used in many places so I don't know if this would be a satisfactory solution for you.

Reply:
I dont not think it is a good idea to disable sanitizer on `type operator()(U u)` either. 

We can conditionally handle floating point to integer conversion inside the convert operator itself. 
```
  shape::visit(type, [&](auto as) {
                // clamping value between target_type's max and min doesn't work for NaNs,
                if(std::isnan(static_cast<double>(x)))
                {
                    y = as.nan();
                }
---------------------> Here 
               // if "type" is integer and "x" has floating point then, first do the clamping and then do the conversion. 
------------------------
                else
                {
                    // clamp overflowing/underflowing values to min()/max() instead of +/-infinity
                    // during downcasting
                    y = std::min(std::max(as(x), as.min()), as.max());
                }
            });

```

Reply:
Doing this works inside the `convert.hpp`'s `apply()` function.

```
  else if(shape::is_integral(type) and std::is_floating_point_v<decltype(x)>)
                {
                    // for the floating point to integer conversion, clamp first and then convert to
                    // avoid undefined behaviour
                    y = as(std::min(std::max(static_cast<double>(x), static_cast<double>(as.min())),
                                    static_cast<double>(as.max())));
                }
```

Reply:
Looks like we can better combine both `else` branches. 

```
 shape::visit(type, [&](auto as) {
                // clamping value between target_type's max and min doesn't work for NaNs,
                if(std::isnan(static_cast<double>(x)))
                {
                    y = as.nan();
                }
                else 
                {
                    // for the floating point to integer conversion, clamp first and then convert to
                    // avoid undefined behaviour.  For the floating point to floating point downcast, clamp first to avoid underflow/overflow to map to -Inf/Inf. 

                    y = as(std::min(std::max(static_cast<double>(x), static_cast<double>(as.min())),
                                    static_cast<double>(as.max())));
                }
            });
```

Reply:
@nives-vukovic, could you please provide simple numerical example(s), of the exact conversion that is/are an issue. @umangyadav, you might have a point in merging the new `else_if` with the existing else clause, however I don't understand the logic of casting to doubles here -- to even detect the NaN in the original `if clause`! 

Is this conversion valid for only fp32<->int or also some fp16 or fp8 variant converting to int?

Reply:
@lakhinderwalia  So I will try to summarize discussion from https://github.com/ROCm/AMDMIGraphX/pull/2973:
Clang asan error has been reported for one maxpool verify test where uint8 is used: /home/jenkins/workspace/AMDMIGraphX_PR-2973/src/include/migraphx/shape.hpp:338:25: runtime error: -3.40282e+38 is outside the range of representable values of type 'unsigned char'. As Umang pointed out there, test is in ceil mode and last window of maxpool is operation on padded region completely which are initialized with numeric_limits<float>::Lowest(). 

As I wrote above from clang documentation: Error that is triggered is float-cast-overflow: Conversion to, from, or between floating-point types which would overflow the destination. Because the range of representable values for all floating-point types supported by Clang is [-inf, +inf], the only cases detected are conversions from floating point to integer types.

Reply:
@lakhinderwalia  https://github.com/ROCm/AMDMIGraphX/pull/2283 Check this, it would answer your question about casting to double for checking NaN.

Reply:
So what @umangyadav wrote is mostly correct. I dont know if there is an issue when converting between large int64 types(like uint64 to int64) as these can represent large integer values that can't be represented in double(so maybe a another `else` branch for this edge case).

Reply:
Pardon me, I am still confused about the actual example. I don't know where are the NaNs coming from. 

I did this simple conversion experiment on my linux m/c (gcc):

```
  float zf = std::numeric_limits<float>::lowest();
  std::cout << "zf convert: " << zf << '\t' << uint32_t(zf) << '\t' << int32_t(zf)
	    << '\t' << short(zf) << "\t char: " << std::hex << (int) uint8_t(zf) << std::endl;

```
And it correctly outputs/converts: 
`zf convert: -3.40282e+38	0	-2147483648	0	 char: 0`

Reply:
Are you building with UBSAN on?

Reply:
It was compiled as : `g++ -fsanitize=address -libasan`

Reply:
> It was compiled as : `g++ -fsanitize=address`

Try adding `-fsanitize=undefined` as well

Reply:
(Yes that option does highlight the issue. Thanks.) 

One answer is: https://clang.llvm.org/docs/UndefinedBehaviorSanitizer.html#disabling-instrumentation-with-attribute-no-sanitize-undefined

Reply:
> (Yes that option does highlight the issue. Thanks.)
> 
> One answer is: https://clang.llvm.org/docs/UndefinedBehaviorSanitizer.html#disabling-instrumentation-with-attribute-no-sanitize-undefined

It has been tried but that would disable sanitizer on entire type cast function in migraphx. it is not desired. Therefore need to handle it inside the convert.hpp itself

Reply:
Okay. Thanks. 
Maybe try, on a test_case basis,  passing on an environment flag/variable: https://clang.llvm.org/docs/UndefinedBehaviorSanitizer.html#id13